File: cpp-src/hex.cpp
/*
    start playing the game:  this is the "main" for running the game


    Run as hex [size] [n_trials]
*/

#include "hex.h"

int main(int argc, char *argv[])
{
    int size = 5;
    int n_trials = 1000;

    try {
        if (argc == 1) {
            ;  // run with defaults
        }
        else if (argc == 2) {
            size = std::stoi(argv[1]);
        }
        else if (argc == 3) {
            size = std::stoi(argv[1]);
            n_trials = std::stoi(argv[2]);
        }
        else {
            cerr << "Wrong number of input arguments:\n"
                << "Run as hex [size] [n_trials]. exiting..." << endl;
            return 1;
        }
    }
    catch (const std::invalid_argument& e) {
        cerr << "Error: Invalid argument. Please provide valid integers for size and n_trials.\n";
        return 1;
    }
    catch (const std::out_of_range& e) {
        cerr << "Error: Argument out of range. Please provide smaller values.\n";
        return 1;
    }

    if (size <= 0) {
        cerr << "Error: Bad size input. Must be a positive integer.\n";
        return 1;
    }

    Hex hb(size);  // create the game object
    hb.make_board();

    hb.play_game(n_trials);

    // cout << "Assessing who won took " << hb.winner_assess_time.show() << " seconds.\n";
    cout << "Simulating and evaluating moves took "
        << hb.move_simulation_time.show() << " seconds.\n";

    return 0;
}

File: nim-src/hex.nim
# nim c -d:release --mm:arc --opt:speed -d:lto --outdir:nim_build nim-src/hex.nim

#######################################################
# script to initiate the game
#######################################################

import 
  os,
  strutils,
  # project modules
  hex_board,
  game_play

proc main() =
  var 
    size: int = 5
    n_trials: int = 1500
    debug: bool = false

  let 
      pcnt: int = paramCount()
      
  if pcnt == 0:
    discard  # do nothing and run with defaults
  elif pcnt == 1:
    size = parseInt(paramStr(1))
  elif pcnt == 2:
    size = parseInt(paramStr(1)) 
    n_trials = parseInt(paramStr(2))
    if (n_trials < 250):
      echo("Number of trials input, 2nd parameter, must be between 250 and 10000:")
      echo("Proceeding with n_trials = 250")
      n_trials = 250
  elif pcnt == 3:
    size = parseInt(paramStr(1)) 
    n_trials = parseInt(paramStr(2))    
    debug = parseBool(paramstr(3))
  else:    
    echo "Wrong number of input arguments."
    echo "Run as hex [size] [n_trials] [debug] Exiting..."
    echo "To enter the 2nd or 3rd arguments, you must also enter the preceding arguments."
    quit()

  if (size < 2 or size > 21) : 
    quit("Bad size input. Must be positive integer between 2 and 21. Exiting...")

  var hb = initHexboard(size) # create the game object
  hb.make_hex_graph()

  hb.play_game(n_trials, debug)

  echo("Simulating and evaluating moves took ", 
    hb.move_sim_time_cum, " seconds.\n");

  if debug:
    hb.save_diagnostics()

when isMainModule:
  main()
